Add AI Fallback System with Retry, Circuit Breaker, and Monitoring #IEEESOC

Satyamgupta2365 opened the pull request:
This PR addresses issue #[2] by implementing a robust fallback system for AI model calls to improve reliability and user experience during service disruptions or rate limiting. The changes introduce a wrapper service for AI calls, fallback models, retry logic with exponential backoff, circuit breaker pattern, graceful degradation, and a monitoring dashboard.
 Learnify is an educational platform leveraging AI for quiz generation, PDF chat, and language tutoring, built with React, Supabase, and Vite.

 ## AI Fallback System

 The project includes a robust AI fallback system to handle service disruptions and rate limiting, ensuring reliability for AI-powered features.

 ### Features
 - **Fallback AI Models**: Uses GROQ as the primary service and OpenAI as fallback for quiz generation, PDF chat, and language tutor features.
 - **Retry Logic**: Implements exponential backoff for retries (configured via `VITE_AI_RETRY_ATTEMPTS`).
 - **Circuit Breaker**: Prevents repeated calls to failing services (`src/services/circuitBreaker.ts`).
 - **Graceful Degradation**: Provides dynamic mock responses (`src/services/mockAI.ts`) when both primary and fallback services fail.
 - **Real-Time Monitoring**: Displays AI service health (feature, provider, duration, status, error) on the admin dashboard (`/admin`) using Supabase Realtime (`src/components/MonitoringDashboard.tsx`).
 - **Access Control**: Restricts the admin dashboard to users with the `admin` role (`src/hooks/useAuth.ts`).

 ### Setup
 1. Clone the repository:
    ```bash
    git clone https://github.com/tarinagarwal/learnify.git
    cd learnify
    ```
 2. Install dependencies:
    ```bash
    npm install
    npm install axios @supabase/supabase-js react-router-dom
    npm install --save-dev jest @testing-library/react @testing-library/jest-dom axios-mock-adapter
    ```
 3. Create a `.env` file with the following:
    ```env
    VITE_SUPABASE_URL=your_supabase_url
    VITE_SUPABASE_ANON_KEY=your_supabase_anon_key
    VITE_GROQ_API_KEY=your_groq_api_key
    VITE_OPENAI_API_KEY=your_openai_api_key
    VITE_AI_RETRY_ATTEMPTS=3
    VITE_AI_CIRCUIT_BREAKER_TIMEOUT=60000
    VITE_AI_CIRCUIT_BREAKER_THRESHOLD=5
    ```
 4. Apply Supabase migrations:
    ```bash
    supabase db push
    ```
 5. Run the development server:
    ```bash
    npm run dev
    ```
 6. Navigate to `http://localhost:5173/admin` to view the AI monitoring dashboard (requires admin role).

 ### Database
 - **ai_metrics**: Stores AI call metrics for the monitoring dashboard.
 - **profiles**: Updated with a `role` column for access control (`ALTER TABLE` in `supabase/migrations/202505191137_alter_profiles.sql`).

 ### Testing
 - Unit tests for `aiService.ts` (`src/services/aiService.test.ts`) verify handling of malformed JSON and non-JSON responses.
 - Run tests with:
    ```bash
    npm test
    ```

 ### Notes
 - Update API endpoints in `src/services/aiService.ts` to match actual GROQ/OpenAI endpoints.
 - Ensure `.env` is not committed (listed in `.gitignore`).
Closes #[2]

tarinagarwal commented:
@Satyamgupta2365 
Does the MonitoringDashboard auto-refresh? If not consider using WebSockets for real time updates.
Also, were any changes made to the database? If yes, consider committing the migration file.
Are you ensuring proper access control on the AdminDashboard page, and have you added it to the routes?
Have you tested scenarios where all services return unexpected results, such as malformed JSON objects?

Satyamgupta2365 commented:
Thanks for the feedback! I’ve addressed all points:

1. **MonitoringDashboard**: Replaced polling with Supabase Realtime (WebSockets) for real-time updates in `src/components/MonitoringDashboard.tsx`.
2. **Database**: Added `ai_metrics` and `profiles` tables with migrations in `supabase/migrations/202505191136_create_ai_metrics.sql` and `202505191137_create_profiles.sql`.
3. **Access Control & Routing**: Protected `AdminDashboard` (`src/pages/AdminDashboard.tsx`) with `useAuth` hook (`src/hooks/useAuth.ts`) for admin-only access. Added `/admin`, `/unauthorized`, and `/login` routes in `App.tsx`.
4. **Error Handling**: Enhanced `aiService.ts` to handle malformed JSON/non-JSON responses, with tests in `aiService.test.ts`.

New files include `aiTypes.ts`, `aiConfig.ts`, `circuitBreaker.ts`, `monitoringService.ts`, `mockAI.ts`, `logger.ts`, and more. Please update `.env` with valid keys and adjust API endpoints in `aiService.ts`. Let me know if further tweaks are needed!

tarinagarwal commented:
Good work @Satyamgupta2365 , but few concerns
- use ALTER TABLE instead of CREATE TABLE for profiles since we already have a table named profiles. 
- Static mock responses only? Should plan to use dynamic or at least parameterized responses.
- Using any type on payload, should implement stricter type from AIResponse, probably a type map

Satyamgupta2365 commented:
Implemented all suggested changes @tarinagarwal 

- Replaced `CREATE TABLE` with `ALTER TABLE` for `profiles`.
- Mock responses now use parameterized data.
- Replaced `any` with stricter types based on `AIResponse`.
- Updated `README.md` with latest changes.

Tested and everything’s working smoothly!

tarinagarwal reviewed (approved):
LGTM!...thanks

tarinagarwal commented:
@Satyamgupta2365 Points will be updated soon
Templates are available on the repo, you can raise some and get assigned to particular tasks

Satyamgupta2365 commented:
Thank you
